controllers/controlplaneupgrade_controller.go

```diff
@@ -234,47 +237,89 @@ func (r *ControlPlaneUpgradeReconciler) updateStatus(ctx context.Context, log lo
+	}
+
+	// Update the machine k8s version and update the KubeadmClusterConfiguration annotation
+	machine.Spec.Version = &nodeUpgrade.Spec.KubernetesVersion
+	annotations.AddAnnotations(machine, map[string]string{kubeadmClusterConfigurationAnnotation: string(kcc)})
 
 	if err := machinePatchHelper.Patch(ctx, machine); err != nil {
 		return fmt.Errorf("updating spec for machine %s: %v", machine.Name, err)
 	}
 
+	if err := r.updateKubeadmConfig(ctx, log, kcpSpec, machine); err != nil {
+		return fmt.Errorf("updating kubeadm config: %v", err)
+	}
+
 	return nil
 }
 
-func getKubeadmClusterConfig(cpUpgrade *anywherev1.ControlPlaneUpgrade) (string, error) {
-	kcpSpec := &controlplanev1.KubeadmControlPlaneSpec{}
-	decodedKcpSpec, err := base64.StdEncoding.DecodeString(cpUpgrade.Spec.ControlPlaneSpecData)
+func (r *ControlPlaneUpgradeReconciler) updateKubeadmConfig(ctx context.Context, log logr.Logger, kcpSpec *controlplanev1.KubeadmControlPlaneSpec, machine *clusterv1.Machine) error {
+	bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+	if bootstrapRef == nil {
+		return fmt.Errorf("bootstrap config for machine %s is nil", machine.Name)
+	}
+
+	kc := &bootstrapv1.KubeadmConfig{}
+	if err := r.client.Get(ctx, types.NamespacedName{Name: bootstrapRef.Name, Namespace: machine.Namespace}, kc); err != nil {
+		return fmt.Errorf("retrieving bootstrap config for machine %s: %v", machine.Name, err)
+	}
+
+	patchHelper, err := patch.NewHelper(kc, r.client)
 	if err != nil {
-		return "", fmt.Errorf("decoding cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+		return fmt.Errorf("initializing patch helper for KubeadmConfig %s: %v", kc.Name, err)
 	}
-	if err := json.Unmarshal(decodedKcpSpec, kcpSpec); err != nil {
-		return "", fmt.Errorf("unmarshaling cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+
+	kcSpec := kcpSpec.KubeadmConfigSpec.DeepCopy()
+	if kc.Spec.InitConfiguration == nil {
+		kcSpec.InitConfiguration = nil
+	}
+
+	if kc.Spec.JoinConfiguration == nil {
+		kcSpec.JoinConfiguration = nil
+	}
+
+	kc.Spec = *kcSpec
+	log.Info("Patching KubeadmConfig", "KubeadmConfig", kc.Name)
+	if err := patchHelper.Patch(ctx, kc); err != nil {
+		return fmt.Errorf("patching KubeadmConfig %s for Machine %s: %v", kc.Name, machine.Name, err)
 	}
-	clusterConfig, err := json.Marshal(kcpSpec.KubeadmConfigSpec.ClusterConfiguration)
+
+	return nil
+}
+
+func decodeAndUnmarshalKcpSpecData(kcpSpecData string) (*controlplanev1.KubeadmControlPlaneSpec, error) {
+	kcpSpec := &controlplanev1.KubeadmControlPlaneSpec{}
+	decodedKcpSpec, err := base64.StdEncoding.DecodeString(kcpSpecData)
 	if err != nil {
-		return "", fmt.Errorf("marshaling KCP cluster configuration: %v", err)
+		return nil, fmt.Errorf("decoding cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+	}
+	if err := json.Unmarshal(decodedKcpSpec, kcpSpec); err != nil {
+		return nil, fmt.Errorf("unmarshaling cpUpgrade.Spec.ControlPlaneSpec: %v", err)
 	}
-	return string(clusterConfig), nil
+	return kcpSpec, err
```

Comment:
Why are we returning err if we know it has to be nil?
```suggestion
	return kcpSpec, nil
```